### ISSUE TITLE
support relative URI references in Location header in Agent

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5462]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5462|
|Type|enhancement|
|Created|2012-01-25 07:46:11Z|

While discussing [ticket:5405 the ticket about putting getPage on top of Agent] with amberite, it came to light that one blocking issue was the differing behavior of `getPage` and `Agent` with respect to relative URI references in the Location header on redirects.  While the ticket doesn't contain this discussion (hey, update the ticket next time, jerks), amberite summarized as follows: `getPage` does kinda what browsers do and follows relative redirects, but `Agent` is more spec-ly correct and instead generates an error, so replacing the implementation of `getPage` with one based on Agent might constitute an incompatible change.

Except, browsers are pretty popular HTTP clients, so we should probably emulate their behavior more often than not.  However, this distinction is fast becoming irrelevant anyway, since the spec is evolving to come into line with what browsers do.  Handily, this means we have a formal statement of what exactly a relative URI should do, rather than just some random observations of browser behavior to guess based on.  As [the HTTPbis semantics draft 18 puts it](http://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-18#section-9.5): 

    When ![the Location header] has the form of a relative reference (![RFC3986], Section 4.2), the final value is computed by resolving it against the effective request URI   (![RFC3986], Section 5).

In other words, what `URLPath.click` does. (`URLPath.click`'s exact adherence to RFC 3986 section 5 is a topic for a different ticket.)

<details><summary>Searchable metadata</summary>

```
trac-id__5462 5462
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__agent agent
time__1327477571000000 1327477571000000
changetime__1370424604000000 1370424604000000
version__None None
owner__amberite amberite
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set status to `closed`|
|-|-|

Duplicate of [#5434](https://github.com/twisted/twisted/issues/5434).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_3"></a>|@jonathanj set status to `reopened`|
|-|-|

As mentioned in the original description, according to the latest [HTTP 1.1 bis draft](https://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-22#section-7.1.2) relative location URIs are allowed behaviour. This ticket was previously marked as a duplicate of a ticket proposing a new redirect agent that behaved like a browser.

I'm re-opening this ticket because relative location URIs should be considered good behaviour and not only browser-like.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_4"></a>|@jonathanj commented|
|-|-|

(In [37885]) Resolve relative URIs in RedirectAgent and BrowserLikeRedirectAgent, refs [#5434](https://github.com/twisted/twisted/issues/5434), [#5462](https://github.com/twisted/twisted/issues/5462).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj set status to `closed`|
|-|-|

(In [38672]) Merge browser-like-redirect-agent-5434-2.

Author: jonathanj
Reviewer: tom.prince, rwall
Fixes: [#5434](https://github.com/twisted/twisted/issues/5434), [#5462](https://github.com/twisted/twisted/issues/5462)

Introduce BrowserLikeRedirectAgent, a RedirectAgent subclass, that handles redirects more like a web browser, specifically it treats HTTP 301 and 302 like HTTP 303 on non-HEAD/GET requests and changes the method to GET before proceeding.

RedirectAgent now also resolves relative URI references in the Location header against the most recent request URI.
